{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Assignment 05\n",
    "\n",
    "# Write a Python script called 'parseGFF.py' that\n",
    "# opens and reads the file, 'watermelon.gff'.\n",
    "# Extract the gene name, store it in a list,\n",
    "# then print an alphabetized list of the genes at the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cox1 exon 1 \n",
      "cox1 \n",
      "cox1 exon 2 \n",
      "similar to chloroplast LSC region from ndhC-trnV spacer to atpB-rbcL spacer \n",
      "trnM(CAT)-cp \n",
      "similar to chloroplast LSC region from psbD to psbC \n",
      "similar to chloroplast rbcL (LSC) \n",
      "similar to chloroplast IR region from rps12 to rps12-rps7 spacer \n",
      "rpl2 exon 1 \n",
      "rpl2 \n",
      "rpl2 exon 2 \n",
      "rps19-1 \n",
      "rps3 exon 1 \n",
      "rps3 \n",
      "rps3 exon 2 \n",
      "rpl16 \n",
      "atp6 \n",
      "inverted repeat \n",
      "sdh3-1 \n",
      "trnQ(TTG)-1 \n",
      "trnG(GCC)-1 \n",
      "trnM(CAT)-cp \n",
      "rps4 \n",
      "nad6 \n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to trnfM-rps14 spacer \n",
      "atp8 \n",
      "cox3 \n",
      "sdh4 \n",
      "nad2 exon 5 \n",
      "nad2 \n",
      "nad2 exon 4 \n",
      "nad2 \n",
      "nad2 exon 3 \n",
      "trnY(GTA) \n",
      "trnN(GTT)-cp \n",
      "trnC(GCA)-2 \n",
      "nad3 \n",
      "rps12 \n",
      "cox2 exon 1 \n",
      "cox2 \n",
      "cox2 exon 2 \n",
      "similar to chloroplast psbC (LSC) \n",
      "nad1 exon 1 \n",
      "atp6 fragment \n",
      "atp9 fragment \n",
      "similar to chloroplast rpoC2 \n",
      "atp1 \n",
      "trnC(GCA)-1 \n",
      "trnE(TTC) \n",
      "trnH(GTG)-cp \n",
      "similar to chloroplast LSC region from rps14 to trnfM-rps14 spacer \n",
      "nad6 fragment \n",
      "similar to chloroplast LSC region from trnL-trnL spacer to ndhC-trnV spacer \n",
      "trnF(GAA)-cp \n",
      "rrn26 \n",
      "trnfM(CAT) \n",
      "similar to chloroplast LSC region from trnL-trnL spacer to rps4 \n",
      "trnC(ACA)-cp \n",
      "trnI(CAT) \n",
      "similar to chloroplast LSC region from psbM-trnD spacer to trnT-psbD spacer \n",
      "trnD(GTC)-cp \n",
      "nad4 exon 1 \n",
      "nad4 \n",
      "nad4 exon 2 \n",
      "nad4 \n",
      "nad4 exon 3 \n",
      "nad4 \n",
      "nad4 exon 4 \n",
      "similar to chloroplast IR region from rps7 to rps12-trnV spacer \n",
      "nad1 exon 5 \n",
      "nad1 \n",
      "matR \n",
      "nad1 exon 4 \n",
      "ccmFn \n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to rps14 \n",
      "rps1 \n",
      "cox2 fragment \n",
      "similar to chloroplast LSC region from psbA to trnH-psbA spacer \n",
      "similar to chloroplast LSC region from petG-trnW spacer to trnP-psaJ spacer \n",
      "trnW(CCA)-cp \n",
      "trnP(TGG)-cp \n",
      "trnK(TTT) \n",
      "ccmB \n",
      "inverted repeat \n",
      "trnG(GCC)-2 \n",
      "trnQ(TTG)-2 \n",
      "sdh3-2 \n",
      "nad9 \n",
      "atp9 \n",
      "nad5 exon 1 \n",
      "nad5 \n",
      "nad5 exon 2 \n",
      "similar to chloroplast LSC region from rpoB-trnC spacer to rpoB \n",
      "ccmFc exon 2 \n",
      "ccmFc \n",
      "ccmFc exon 1 \n",
      "nad5 exon 3 \n",
      "atp9 fragment \n",
      "rpl5 \n",
      "cob \n",
      "rps7 \n",
      "trnP(TGG) \n",
      "trnF(GAA) \n",
      "trnS(GCT) \n",
      "rrn5 \n",
      "rrn18 \n",
      "similar to chloroplast rpl32-trnL spacer (SSC) \n",
      "mttB \n",
      "similar to chloroplast IR/LSC region from rpl2 to trnH-psbA spacer \n",
      "trnH(GTG)-cp \n",
      "similar to chloroplast LSC region from atpF-atpH spacer to rpoC2 \n",
      "matR fragment \n",
      "rps13 \n",
      "nad1 exon 2 \n",
      "nad1 \n",
      "nad1 exon 3 \n",
      "sdh3 fragment \n",
      "nad2 exon 1 \n",
      "nad2 \n",
      "nad2 exon 2 \n",
      "similar to chloroplast ycf2 \n",
      "rps19 fragment \n",
      "nad5 exon 4 \n",
      "nad5 \n",
      "nad5 exon 5 \n",
      "nad7 exon 5 \n",
      "nad7 \n",
      "nad7 exon 4 \n",
      "nad7 \n",
      "nad7 exon 3 \n",
      "nad7 \n",
      "nad7 exon 2 \n",
      "nad7 \n",
      "nad7 exon 1 \n",
      "trnS(TGA) \n",
      "t-element for generation of 3' mRNA terminus of ccmC gene, see PubMed ID 17488843, similar to chloroplast-like trnI(TAT), with expanded anticodon loop \n",
      "ccmC \n",
      "atp4 \n",
      "nad4L \n",
      "sdh3 fragment \n",
      "rps19-2 \n",
      "rps10 exon 1 \n",
      "rps10 \n",
      "rps10 exon 2 \n"
     ]
    }
   ],
   "source": [
    "input = open(\"watermelon.gff\", \"r\")\n",
    "\n",
    "for line in input:\n",
    "    value = line.split(\"\\t\", 8) # Split by tab\n",
    "    value = value[8] # Only save 9th column\n",
    "    value = value.split(\";\", 5) # Split by semicolon\n",
    "    value = value[0] # Only save 1st column\n",
    "    value = value.split(\" \", 1) # Split by space\n",
    "    value = value[1] # Only save 2nd column\n",
    "    print(value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "### GRAVEYARD OF FAILED ATTEMPTS\n",
    "\n",
    "# Load modules\n",
    "#import gffutils\n",
    "#import os\n",
    "#import sys\n",
    "#import pandas\n",
    "#from bioinfo_tools.parsers.gff import Gff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change working directory\n",
    "#os.getcwd()\n",
    "#\n",
    "#os.chdir(\"~/Users/simontye/Documents/Research/Courses/Programming)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cox1 exon 1 \n",
      "cox1 \n",
      "cox1 exon 2 \n",
      "similar to chloroplast LSC region from ndhC-trnV spacer to atpB-rbcL spacer \n",
      "trnM(CAT)-cp \n",
      "similar to chloroplast LSC region from psbD to psbC \n",
      "similar to chloroplast rbcL (LSC) \n",
      "similar to chloroplast IR region from rps12 to rps12-rps7 spacer \n",
      "rpl2 exon 1 \n",
      "rpl2 \n",
      "rpl2 exon 2 \n",
      "rps19-1 \n",
      "rps3 exon 1 \n",
      "rps3 \n",
      "rps3 exon 2 \n",
      "rpl16 \n",
      "atp6 \n",
      "inverted repeat \n",
      "sdh3-1 \n",
      "trnQ(TTG)-1 \n",
      "trnG(GCC)-1 \n",
      "trnM(CAT)-cp \n",
      "rps4 \n",
      "nad6 \n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to trnfM-rps14 spacer \n",
      "atp8 \n",
      "cox3 \n",
      "sdh4 \n",
      "nad2 exon 5 \n",
      "nad2 \n",
      "nad2 exon 4 \n",
      "nad2 \n",
      "nad2 exon 3 \n",
      "trnY(GTA) \n",
      "trnN(GTT)-cp \n",
      "trnC(GCA)-2 \n",
      "nad3 \n",
      "rps12 \n",
      "cox2 exon 1 \n",
      "cox2 \n",
      "cox2 exon 2 \n",
      "similar to chloroplast psbC (LSC) \n",
      "nad1 exon 1 \n",
      "atp6 fragment \n",
      "atp9 fragment \n",
      "similar to chloroplast rpoC2 \n",
      "atp1 \n",
      "trnC(GCA)-1 \n",
      "trnE(TTC) \n",
      "trnH(GTG)-cp \n",
      "similar to chloroplast LSC region from rps14 to trnfM-rps14 spacer \n",
      "nad6 fragment \n",
      "similar to chloroplast LSC region from trnL-trnL spacer to ndhC-trnV spacer \n",
      "trnF(GAA)-cp \n",
      "rrn26 \n",
      "trnfM(CAT) \n",
      "similar to chloroplast LSC region from trnL-trnL spacer to rps4 \n",
      "trnC(ACA)-cp \n",
      "trnI(CAT) \n",
      "similar to chloroplast LSC region from psbM-trnD spacer to trnT-psbD spacer \n",
      "trnD(GTC)-cp \n",
      "nad4 exon 1 \n",
      "nad4 \n",
      "nad4 exon 2 \n",
      "nad4 \n",
      "nad4 exon 3 \n",
      "nad4 \n",
      "nad4 exon 4 \n",
      "similar to chloroplast IR region from rps7 to rps12-trnV spacer \n",
      "nad1 exon 5 \n",
      "nad1 \n",
      "matR \n",
      "nad1 exon 4 \n",
      "ccmFn \n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to rps14 \n",
      "rps1 \n",
      "cox2 fragment \n",
      "similar to chloroplast LSC region from psbA to trnH-psbA spacer \n",
      "similar to chloroplast LSC region from petG-trnW spacer to trnP-psaJ spacer \n",
      "trnW(CCA)-cp \n",
      "trnP(TGG)-cp \n",
      "trnK(TTT) \n",
      "ccmB \n",
      "inverted repeat \n",
      "trnG(GCC)-2 \n",
      "trnQ(TTG)-2 \n",
      "sdh3-2 \n",
      "nad9 \n",
      "atp9 \n",
      "nad5 exon 1 \n",
      "nad5 \n",
      "nad5 exon 2 \n",
      "similar to chloroplast LSC region from rpoB-trnC spacer to rpoB \n",
      "ccmFc exon 2 \n",
      "ccmFc \n",
      "ccmFc exon 1 \n",
      "nad5 exon 3 \n",
      "atp9 fragment \n",
      "rpl5 \n",
      "cob \n",
      "rps7 \n",
      "trnP(TGG) \n",
      "trnF(GAA) \n",
      "trnS(GCT) \n",
      "rrn5 \n",
      "rrn18 \n",
      "similar to chloroplast rpl32-trnL spacer (SSC) \n",
      "mttB \n",
      "similar to chloroplast IR/LSC region from rpl2 to trnH-psbA spacer \n",
      "trnH(GTG)-cp \n",
      "similar to chloroplast LSC region from atpF-atpH spacer to rpoC2 \n",
      "matR fragment \n",
      "rps13 \n",
      "nad1 exon 2 \n",
      "nad1 \n",
      "nad1 exon 3 \n",
      "sdh3 fragment \n",
      "nad2 exon 1 \n",
      "nad2 \n",
      "nad2 exon 2 \n",
      "similar to chloroplast ycf2 \n",
      "rps19 fragment \n",
      "nad5 exon 4 \n",
      "nad5 \n",
      "nad5 exon 5 \n",
      "nad7 exon 5 \n",
      "nad7 \n",
      "nad7 exon 4 \n",
      "nad7 \n",
      "nad7 exon 3 \n",
      "nad7 \n",
      "nad7 exon 2 \n",
      "nad7 \n",
      "nad7 exon 1 \n",
      "trnS(TGA) \n",
      "t-element for generation of 3' mRNA terminus of ccmC gene, see PubMed ID 17488843, similar to chloroplast-like trnI(TAT), with expanded anticodon loop \n",
      "ccmC \n",
      "atp4 \n",
      "nad4L \n",
      "sdh3 fragment \n",
      "rps19-2 \n",
      "rps10 exon 1 \n",
      "rps10 \n",
      "rps10 exon 2 \n"
     ]
    }
   ],
   "source": [
    "# Load file\n",
    "#file = gffutils.example_filename(\"/Users/simontye/Documents/Research/Courses/Programming/Assignments/watermelon.gff\")\n",
    "\n",
    "# Open, read, and print file\n",
    "#print(open(file).read()) \n",
    "\n",
    "# Create a local sqlite3 file-based database\n",
    "#db = gffutils.create_db(file, dbfn='test.db', force = True, keep_order = True, merge_strategy = 'merge', sort_attribute_values = True)\n",
    "\n",
    "# Connect to database\n",
    "#db = gffutils.FeatureDB('test.db', keep_order = True)\n",
    "    \n",
    "# Details about each gene are retained, but can be removed\n",
    "# by adjusting last two lines of code\n",
    "\n",
    "#for line in input:\n",
    "#    value = line.split(\"\\t\", 8) # Split by tab and select 8th column\n",
    "#    value[8]:\n",
    "#        final.append(value[8])\n",
    "#        x = final.strip(\"\\n\")\n",
    "#        print(x)\n",
    "\n",
    "#input.close()\n",
    "#print(final)\n",
    "\n",
    "#final2 = [remove[6:] for remove in final]\n",
    "\n",
    "        #value = value[1].rstrip()\n",
    "        #value = value.strip('\\\"')\n",
    "        #if value in GeneList:\n",
    "        #    print(value, \"already found.\")\n",
    "        #else:\n",
    "        #    GeneList.append(value)\n",
    "        #    print(value, \"added to list.\")\n",
    "            \n",
    "            \n",
    "# For loop to list gene names\n",
    "#for genewilder in db:\n",
    "#    print(genewilder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'BCBio'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-245-8003140a7cfa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# Load modules\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpprint\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mBCBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mGFFExaminer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;31m#in_file = \"watermelon.gff\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'BCBio'"
     ]
    }
   ],
   "source": [
    "# Install BCBio pack (bash)\n",
    "#git clone git://github.com/chapmanb/bcbb.git\n",
    "#cd bcbb/gff\n",
    "#python setup.py build\n",
    "#sudo python setup.py install\n",
    "\n",
    "# Load modules\n",
    "#import pprint\n",
    "#from BCBio import GFFExaminer\n",
    "\n",
    "#in_file = \"watermelon.gff\"\n",
    "#examiner = GFFExaminer()\n",
    "#in_handle = open(in_file)\n",
    "#pprint.pprint(examiner.parent_child_map(in_handle))\n",
    "#in_handle.close()\n",
    "\n",
    "#########################\n",
    "\n",
    "#from Bio import SeqIO\n",
    "#\n",
    "#for seq_record in SeqIO.parse(\"watermelon.gff\", \"gff\"): print(seq_record.id)\n",
    "#print(repr(seq_record.seq))\n",
    "#print(len(seq_record))\n",
    "\n",
    "#########################\n",
    "\n",
    "\n",
    "## Open watermelon.gff\n",
    "#watermelon = open(\"watermelon.gff\", \"r\")\n",
    "#water = watermelon.read()\n",
    "#print(water)\n",
    "#\n",
    "# Create empty list\n",
    "#genes = []\n",
    "\n",
    "# Extract gene names from watermelon and store in a list\n",
    "#for coronavirus in watermelon:\n",
    "#    genes.append(coronavirus.rstrip())\n",
    "#\n",
    "## For loop\n",
    "#for whatever in genes:\n",
    "#    print(whatever)\n",
    "#    \n",
    "# Close the file\n",
    "#watermelon.close()\n",
    "#\n",
    "# watermelon = open(watermelon, \"r\")\n",
    "# water = watermelon.read()\n",
    "# print(water)\n",
    "#\n",
    "# Print alphabetized list of gene names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#gff_parser = Gff3()\n",
    "#with open(\"/Users/simontye/Documents/Research/Courses/Programming/Assignments/watermelon.gff\", \"r\") as fh:\n",
    "#    for gene in gff_parser.read(fh):\n",
    "#        print(gene)\n",
    "        \n",
    "## Test\n",
    "#for line in sys.stdin:\n",
    "#    convertedLine = \"\"\n",
    "#    chomped_line = line.rstrip(os.linesep)\n",
    "#    if chomped_line.startswith('##'):\n",
    "#        pass\n",
    "#    elif chomped_line.startswith('track'):\n",
    "#        pass\n",
    "#    else:\n",
    "#        elems = chomped_line.split('\\t')\n",
    "#        cols = dict()\n",
    "#        try:\n",
    "#            cols['seqname'] = elems[0].lstrip(' ')\n",
    "#            cols['source'] = elems[1]\n",
    "#            cols['feature'] = elems[2]\n",
    "#            cols['start'] = int(elems[3])\n",
    "#            cols['end'] = int(elems[4])\n",
    "#            cols['score'] = elems[5]\n",
    "#            cols['strand'] = elems[6]\n",
    "#            cols['frame'] = elems[7]\n",
    "#            cols['attributes'] = elems[8].rstrip(' ')\n",
    "#        except IndexError as ie:\n",
    "#            sys.stderr.write(\"[%s] - Error: Input appears to be missing GTF-specific fields (check that your input data is GTF-formatted)\\n\" % (sys.argv[0]))\n",
    "#            sys.exit(os.EX_DATAERR)\n",
    "#        try:\n",
    "#            cols['comments'] = elems[9]\n",
    "#        except IndexError as ie:\n",
    "#            cols['comments'] = None\n",
    "#        attributes = dict(item.strip().split(' ') for item in cols['attributes'].split(';') if item):\n",
    "#            for line in file:\n",
    "#                print(attributes])\n",
    "#\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#for line in input:\n",
    "#    value = line.split(\"\\t\", 8) # Split by tab and select 8th column\n",
    "#    value[8]:\n",
    "#        final.append(value[8])\n",
    "#        x = final.strip(\"\\n\")\n",
    "#        print(x)\n",
    "\n",
    "#input.close()\n",
    "#print(final)\n",
    "\n",
    "#final2 = [remove[6:] for remove in final]\n",
    "\n",
    "        #value = value[1].rstrip()\n",
    "        #value = value.strip('\\\"')\n",
    "        #if value in GeneList:\n",
    "        #    print(value, \"already found.\")\n",
    "        #else:\n",
    "        #    GeneList.append(value)\n",
    "        #    print(value, \"added to list.\")\n",
    "            \n",
    "            \n",
    "# For loop to list gene names\n",
    "#for genewilder in db:\n",
    "#    print(genewilder)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
