{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello world\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] gff fasta\n",
      "ipykernel_launcher.py: error: the following arguments are required: fasta\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    }
   ],
   "source": [
    "#! /usr/bin/env python3\n",
    "\n",
    "### Notes\n",
    "## Single vs. multiple records\n",
    "# SeqIO.parse (multiple records)\n",
    "# SeqIO.read (single recrods)\n",
    "\n",
    "## Print methods or functions or a given variable type\n",
    "# print(dir(genome))\n",
    "\n",
    "## Tests\n",
    "# print(genome.description)\n",
    "# print(len(genome.seq))\n",
    "# print(genome.seq)\n",
    "\n",
    "# Import modules\n",
    "import argparse\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Create argument parser object\n",
    "parser = argparse.ArgumentParser(description = \"This script parses a GFF file\")\n",
    "\n",
    "# Add positional arguments\n",
    "parser.add_argument(\"gff\",   help = \"Name of the GFF file\")\n",
    "parser.add_argument(\"fasta\", help = \"Name of the FASTA file\")\n",
    "\n",
    "# Parse the arguments\n",
    "args = parser.parse_args()\n",
    "\n",
    "# Read and parse the fasta file\n",
    "genome = SeqIO.read(args.fasta, \"fasta\")\n",
    "genome_seq = genome.seq\n",
    "\n",
    "# Create function\n",
    "with open(args.gff, \"r\") as gff_file:\n",
    "    reader = csv.reader(gff_file, delimiter = \"\\t\")\n",
    "    for line in reader:\n",
    "        if line[2] != \"CDS\":\n",
    "            # Define start and end positions\n",
    "            start = line[3]\n",
    "            end   = line[4]\n",
    "            # Find substrings in genome_seq\n",
    "            substring = genome_seq[(start - 1):(end - 1)]\n",
    "            # Calculate GC content\n",
    "            g_count = substring.count(\"G\")\n",
    "            c_count = substring.count(\"C\")\n",
    "            length  = len(substring)\n",
    "            # Print output\n",
    "            print((g_count + c_count)/length)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atp1\n",
      "atp4\n",
      "atp6\n",
      "atp6 fragment\n",
      "atp8\n",
      "atp9\n",
      "atp9 fragment\n",
      "atp9 fragment\n",
      "ccmB\n",
      "ccmC\n",
      "ccmFc\n",
      "ccmFc exon 1\n",
      "ccmFc exon 2\n",
      "ccmFn\n",
      "cob\n",
      "cox1\n",
      "cox1 exon 1\n",
      "cox1 exon 2\n",
      "cox2\n",
      "cox2 exon 1\n",
      "cox2 exon 2\n",
      "cox2 fragment\n",
      "cox3\n",
      "inverted repeat\n",
      "inverted repeat\n",
      "matR\n",
      "matR fragment\n",
      "mttB\n",
      "nad1\n",
      "nad1\n",
      "nad1 exon 1\n",
      "nad1 exon 2\n",
      "nad1 exon 3\n",
      "nad1 exon 4\n",
      "nad1 exon 5\n",
      "nad2\n",
      "nad2\n",
      "nad2\n",
      "nad2 exon 1\n",
      "nad2 exon 2\n",
      "nad2 exon 3\n",
      "nad2 exon 4\n",
      "nad2 exon 5\n",
      "nad3\n",
      "nad4\n",
      "nad4\n",
      "nad4\n",
      "nad4 exon 1\n",
      "nad4 exon 2\n",
      "nad4 exon 3\n",
      "nad4 exon 4\n",
      "nad4L\n",
      "nad5\n",
      "nad5\n",
      "nad5 exon 1\n",
      "nad5 exon 2\n",
      "nad5 exon 3\n",
      "nad5 exon 4\n",
      "nad5 exon 5\n",
      "nad6\n",
      "nad6 fragment\n",
      "nad7\n",
      "nad7\n",
      "nad7\n",
      "nad7\n",
      "nad7 exon 1\n",
      "nad7 exon 2\n",
      "nad7 exon 3\n",
      "nad7 exon 4\n",
      "nad7 exon 5\n",
      "nad9\n",
      "rpl16\n",
      "rpl2\n",
      "rpl2 exon 1\n",
      "rpl2 exon 2\n",
      "rpl5\n",
      "rps1\n",
      "rps10\n",
      "rps10 exon 1\n",
      "rps10 exon 2\n",
      "rps12\n",
      "rps13\n",
      "rps19 fragment\n",
      "rps19-1\n",
      "rps19-2\n",
      "rps3\n",
      "rps3 exon 1\n",
      "rps3 exon 2\n",
      "rps4\n",
      "rps7\n",
      "rrn18\n",
      "rrn26\n",
      "rrn5\n",
      "sdh3 fragment\n",
      "sdh3 fragment\n",
      "sdh3-1\n",
      "sdh3-2\n",
      "sdh4\n",
      "similar to chloroplast IR region from rps12 to rps12-rps7 spacer\n",
      "similar to chloroplast IR region from rps7 to rps12-trnV spacer\n",
      "similar to chloroplast IR/LSC region from rpl2 to trnH-psbA spacer\n",
      "similar to chloroplast LSC region from atpF-atpH spacer to rpoC2\n",
      "similar to chloroplast LSC region from ndhC-trnV spacer to atpB-rbcL spacer\n",
      "similar to chloroplast LSC region from petG-trnW spacer to trnP-psaJ spacer\n",
      "similar to chloroplast LSC region from psbA to trnH-psbA spacer\n",
      "similar to chloroplast LSC region from psbD to psbC\n",
      "similar to chloroplast LSC region from psbM-trnD spacer to trnT-psbD spacer\n",
      "similar to chloroplast LSC region from rpoB-trnC spacer to rpoB\n",
      "similar to chloroplast LSC region from rps14 to trnfM-rps14 spacer\n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to rps14\n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to trnfM-rps14 spacer\n",
      "similar to chloroplast LSC region from trnL-trnL spacer to ndhC-trnV spacer\n",
      "similar to chloroplast LSC region from trnL-trnL spacer to rps4\n",
      "similar to chloroplast psbC (LSC)\n",
      "similar to chloroplast rbcL (LSC)\n",
      "similar to chloroplast rpl32-trnL spacer (SSC)\n",
      "similar to chloroplast rpoC2\n",
      "similar to chloroplast ycf2\n",
      "t-element for generation of 3' mRNA terminus of ccmC gene, see PubMed ID 17488843, similar to chloroplast-like trnI(TAT), with expanded anticodon loop\n",
      "trnC(ACA)-cp\n",
      "trnC(GCA)-1\n",
      "trnC(GCA)-2\n",
      "trnD(GTC)-cp\n",
      "trnE(TTC)\n",
      "trnF(GAA)\n",
      "trnF(GAA)-cp\n",
      "trnG(GCC)-1\n",
      "trnG(GCC)-2\n",
      "trnH(GTG)-cp\n",
      "trnH(GTG)-cp\n",
      "trnI(CAT)\n",
      "trnK(TTT)\n",
      "trnM(CAT)-cp\n",
      "trnM(CAT)-cp\n",
      "trnN(GTT)-cp\n",
      "trnP(TGG)\n",
      "trnP(TGG)-cp\n",
      "trnQ(TTG)-1\n",
      "trnQ(TTG)-2\n",
      "trnS(GCT)\n",
      "trnS(TGA)\n",
      "trnW(CCA)-cp\n",
      "trnY(GTA)\n",
      "trnfM(CAT)\n"
     ]
    }
   ],
   "source": [
    "### Andy's parseGFF.py\n",
    "\n",
    "# specify the input files\n",
    "gff_file   = 'watermelon.gff'\n",
    "fasta_file = 'watermelon.fsa'\n",
    "\n",
    "\n",
    "# open the GFF file\n",
    "gff = open(gff_file, 'r')\n",
    "\n",
    "\n",
    "# list for gene names\n",
    "gene_names = []\n",
    "\n",
    "# read GFF file, line by line\n",
    "for line in gff:\n",
    "\n",
    "    # remove line breaks\n",
    "    line = line.rstrip('\\n')\n",
    "    \n",
    "    # split each line on the tab character\n",
    "    # sequence, source, feature, begin, end, length, strand, phase, attributes = line.split('\\t')\n",
    "    fields = line.split('\\t')\n",
    "\n",
    "    # split the attributes field (index = 8), each attribute is separated by ' ; '\n",
    "\n",
    "    attributes = fields[8].split(' ; ')\n",
    "\n",
    "    # we only need the first item in this list\n",
    "    # print(attributes[0])\n",
    "    gene_fields = attributes[0].split('Gene ')\n",
    "\n",
    "    # print the gene name\n",
    "    # print(gene_fields[1])\n",
    "\n",
    "    # store the gene name\n",
    "    gene_names.append(gene_fields[1])\n",
    "\n",
    "\n",
    "    # extract the DNA sequence from the genome for this feature\n",
    "\n",
    "    # print the DNA sequence for this feature\n",
    "\n",
    "    # calculate the GC content for this feature, and print it to the screen\n",
    "\n",
    "# print the genes in alphabetical order\n",
    "for gene in sorted(gene_names):\n",
    "    print(gene)\n",
    "\n",
    "gff.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cox1 exon 1 \n",
      "cox1 \n",
      "cox1 exon 2 \n",
      "similar to chloroplast LSC region from ndhC-trnV spacer to atpB-rbcL spacer \n",
      "trnM(CAT)-cp \n",
      "similar to chloroplast LSC region from psbD to psbC \n",
      "similar to chloroplast rbcL (LSC) \n",
      "similar to chloroplast IR region from rps12 to rps12-rps7 spacer \n",
      "rpl2 exon 1 \n",
      "rpl2 \n",
      "rpl2 exon 2 \n",
      "rps19-1 \n",
      "rps3 exon 1 \n",
      "rps3 \n",
      "rps3 exon 2 \n",
      "rpl16 \n",
      "atp6 \n",
      "inverted repeat \n",
      "sdh3-1 \n",
      "trnQ(TTG)-1 \n",
      "trnG(GCC)-1 \n",
      "trnM(CAT)-cp \n",
      "rps4 \n",
      "nad6 \n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to trnfM-rps14 spacer \n",
      "atp8 \n",
      "cox3 \n",
      "sdh4 \n",
      "nad2 exon 5 \n",
      "nad2 \n",
      "nad2 exon 4 \n",
      "nad2 \n",
      "nad2 exon 3 \n",
      "trnY(GTA) \n",
      "trnN(GTT)-cp \n",
      "trnC(GCA)-2 \n",
      "nad3 \n",
      "rps12 \n",
      "cox2 exon 1 \n",
      "cox2 \n",
      "cox2 exon 2 \n",
      "similar to chloroplast psbC (LSC) \n",
      "nad1 exon 1 \n",
      "atp6 fragment \n",
      "atp9 fragment \n",
      "similar to chloroplast rpoC2 \n",
      "atp1 \n",
      "trnC(GCA)-1 \n",
      "trnE(TTC) \n",
      "trnH(GTG)-cp \n",
      "similar to chloroplast LSC region from rps14 to trnfM-rps14 spacer \n",
      "nad6 fragment \n",
      "similar to chloroplast LSC region from trnL-trnL spacer to ndhC-trnV spacer \n",
      "trnF(GAA)-cp \n",
      "rrn26 \n",
      "trnfM(CAT) \n",
      "similar to chloroplast LSC region from trnL-trnL spacer to rps4 \n",
      "trnC(ACA)-cp \n",
      "trnI(CAT) \n",
      "similar to chloroplast LSC region from psbM-trnD spacer to trnT-psbD spacer \n",
      "trnD(GTC)-cp \n",
      "nad4 exon 1 \n",
      "nad4 \n",
      "nad4 exon 2 \n",
      "nad4 \n",
      "nad4 exon 3 \n",
      "nad4 \n",
      "nad4 exon 4 \n",
      "similar to chloroplast IR region from rps7 to rps12-trnV spacer \n",
      "nad1 exon 5 \n",
      "nad1 \n",
      "matR \n",
      "nad1 exon 4 \n",
      "ccmFn \n",
      "similar to chloroplast LSC region from trnG-trnfM spacer to rps14 \n",
      "rps1 \n",
      "cox2 fragment \n",
      "similar to chloroplast LSC region from psbA to trnH-psbA spacer \n",
      "similar to chloroplast LSC region from petG-trnW spacer to trnP-psaJ spacer \n",
      "trnW(CCA)-cp \n",
      "trnP(TGG)-cp \n",
      "trnK(TTT) \n",
      "ccmB \n",
      "inverted repeat \n",
      "trnG(GCC)-2 \n",
      "trnQ(TTG)-2 \n",
      "sdh3-2 \n",
      "nad9 \n",
      "atp9 \n",
      "nad5 exon 1 \n",
      "nad5 \n",
      "nad5 exon 2 \n",
      "similar to chloroplast LSC region from rpoB-trnC spacer to rpoB \n",
      "ccmFc exon 2 \n",
      "ccmFc \n",
      "ccmFc exon 1 \n",
      "nad5 exon 3 \n",
      "atp9 fragment \n",
      "rpl5 \n",
      "cob \n",
      "rps7 \n",
      "trnP(TGG) \n",
      "trnF(GAA) \n",
      "trnS(GCT) \n",
      "rrn5 \n",
      "rrn18 \n",
      "similar to chloroplast rpl32-trnL spacer (SSC) \n",
      "mttB \n",
      "similar to chloroplast IR/LSC region from rpl2 to trnH-psbA spacer \n",
      "trnH(GTG)-cp \n",
      "similar to chloroplast LSC region from atpF-atpH spacer to rpoC2 \n",
      "matR fragment \n",
      "rps13 \n",
      "nad1 exon 2 \n",
      "nad1 \n",
      "nad1 exon 3 \n",
      "sdh3 fragment \n",
      "nad2 exon 1 \n",
      "nad2 \n",
      "nad2 exon 2 \n",
      "similar to chloroplast ycf2 \n",
      "rps19 fragment \n",
      "nad5 exon 4 \n",
      "nad5 \n",
      "nad5 exon 5 \n",
      "nad7 exon 5 \n",
      "nad7 \n",
      "nad7 exon 4 \n",
      "nad7 \n",
      "nad7 exon 3 \n",
      "nad7 \n",
      "nad7 exon 2 \n",
      "nad7 \n",
      "nad7 exon 1 \n",
      "trnS(TGA) \n",
      "t-element for generation of 3' mRNA terminus of ccmC gene, see PubMed ID 17488843, similar to chloroplast-like trnI(TAT), with expanded anticodon loop \n",
      "ccmC \n",
      "atp4 \n",
      "nad4L \n",
      "sdh3 fragment \n",
      "rps19-2 \n",
      "rps10 exon 1 \n",
      "rps10 \n",
      "rps10 exon 2 \n"
     ]
    }
   ],
   "source": [
    "### Simon's parseGFF.py\n",
    "\n",
    "input = open(\"watermelon.gff\", \"r\")\n",
    "\n",
    "for line in input:\n",
    "    value = line.split(\"\\t\", 8) # Split by tab\n",
    "    value = value[8] # Only save 9th column\n",
    "    value = value.split(\";\", 5) # Split by semicolon\n",
    "    value = value[0] # Only save 1st column\n",
    "    value = value.split(\" \", 1) # Split by space\n",
    "    value = value[1] # Only save 2nd column\n",
    "    print(value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Class notes\n",
    "\n",
    "## Updated parse_shaver.py\n",
    "\n",
    "# import module\n",
    "import csv\n",
    "\n",
    "# define the input file\n",
    "infile = \"shaver_etal.csv\"\n",
    "\n",
    "# open the output file\n",
    "with open(\"shaver_tabbed.txt\", \"w\") as output:\n",
    "    # open and parse shaver et al.\n",
    "    with open(infile, 'r') as shaver:\n",
    "        # create a csv reader object\n",
    "        reader = csv.reader(shaver, delimiter = \",\")\n",
    "        for line in reader:\n",
    "        # skip blank lines\n",
    "            if not line:\n",
    "                continue\n",
    "            else:\n",
    "                linewriter = csv.writer(output, delimiter = \"\\t\", quotechar = \"'\")\n",
    "                linewriter.writerow(line)\n",
    "                \n",
    "                \n",
    "                # Print output\n",
    "                #if(line[3] == 'Stepps'):\n",
    "                    #print('\\t'.join([line[3], line[7], line[8]]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'int' is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-f9bbbceb9836>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;31m# add positional argument for the input position in the Fibonacci sequence\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_argument\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"num\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhelp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"The Fibonacci number you wish to calculate\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"int\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# add some optional arguments\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/anaconda3/lib/python3.7/argparse.py\u001b[0m in \u001b[0;36madd_argument\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1362\u001b[0m         \u001b[0mtype_func\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_registry_get\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'type'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maction\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maction\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1363\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype_func\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1364\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'%r is not callable'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtype_func\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1365\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1366\u001b[0m         \u001b[0;31m# raise an error if the metavar does not match the type\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: 'int' is not callable"
     ]
    }
   ],
   "source": [
    "### Class notes\n",
    "\n",
    "## Create an arguement parser\n",
    "\n",
    "# import module\n",
    "import argparse\n",
    "\n",
    "# create an argument parser object\n",
    "parser = argparse.ArgumentParser(description = \"This script returns the Fibonacci number at a specified position in the Fibonacci sequence\")\n",
    "\n",
    "# add positional argument for the input position in the Fibonacci sequence\n",
    "parser.add_argument(\"num\", help = \"The Fibonacci number you wish to calculate\", type = \"int\")\n",
    "# position of positional argument matters\n",
    "parser.add_argument(\"name\", help = \"Name of the user\", type = \"int\")\n",
    "\n",
    "\n",
    "# add some optional arguments\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"-v\", \"--verbose\", action = \"store_true\", help = \"print verbose output\")\n",
    "group.add_argument(\"-s\", \"--simple\", action = \"store_true\", help = \"print simple output (default)\")\n",
    "\n",
    "# parse the arguments\n",
    "args = parser.parse_args()\n",
    "\n",
    "# calculate the Fibonacci number\n",
    "a,b = 0,1\n",
    "\n",
    "for i in range(args.num):\n",
    "    a,b = b, a+b\n",
    "\n",
    "# print the output as requested by the user\n",
    "if(args.verbose):\n",
    "    # print the verbose output\n",
    "    print(\"hello\", args.name, \": for position\", str(args.num), \"the Fibonacci number is \", a)\n",
    "else:\n",
    "    # print the simple output\n",
    "    print(str(args.num), a)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
